### ISSUE TITLE
Valid placeholder in --term is rejected with invalid if there is a placeholder after {cmdline@} with -I

### ISSUE THREAD

**Issue by Parranoh** [enkore/j4-dmenu-desktop#179]
I wanted to save myself a process creation and came up with a custom --term, but it turns out this makes j4dd give the error at [CMDLineTerm.cc:374](https://github.com/enkore/j4-dmenu-desktop/blob/r3.1/src/CMDLineTerm.cc#L374), which should not happen, at least according to the comment there.

Steps to reproduce MWE:

1. Run `j4-dmenu-desktop -I --term-mode custom --term 'alacritty -e {cmdline@} {name}'`
2. Select a terminal application.
3. Observe.

I know this is not a reasonable --term.  I originally wanted something like `alacritty msg create-window -T {name} {cmdline@} || alacritty -T {name} {cmdline@}`.

According to my crude tests, the problem occurs only with `I3Executable` but not with other executables.

**Comment by meator** [comment]
Thanks for reporting! I will investigate this.

~I should also note that~

```
j4-dmenu-desktop --term-mode alacritty
```

~should do what you want, there is no need to set a custom term mode in this case.~

I now understand what you're trying to do. Terminal emulators run in custom term mode are **not** run by a shell. This means that your `||` will not work as you expect it to. A solution to this problem would be to create a shell script with the following contents:

```sh
#!/bin/sh
title="$1"
shift
alacritty msg create-window -T "$title" -e "$@" || alacritty -T "$title" -e "$@"
```

and then call `j4-dmenu-desktop` with

```
j4-dmenu-desktop --term-mode custom --term "/path/to/the/shell/script.sh {name} {cmdline@}"
```

another solution would be to do this:

```
j4-dmenu-desktop --term-mode custom --term '/bin/sh -c alacritty\ msg\ create-window\ -T\ {name}\ -e\ {cmdline*}\ ||\ alacritty\ -T\ {name}\ -e\ {cmdline*}'
```

I would personally prefer the first solution, but the custom term mechanism should be capable enough to make this work without a custom shell script if you really want to.

**Comment by meator** [comment]
This issue should now be fixed in the `develop` branch. @Parranoh, would you mind testing it and verifying that the issue is resolved?

I can provide you instructions on how to build j4-dmenu-desktop from `develop` if you want to.

**Comment by Parranoh** [comment]
Thank you for the quick response!  j4dd now accepts the custom --term `alacritty -e {cmdline@} {name}`, yes, both with `-I` and without.

However the behavior is not quite the same.  If I omit `-I` and launch Htop, as expected, j4dd runs `alacritty -e htop Htop`, that is `htop` with the first argument `Htop`.  If I specify `-I` and launch Htop, it runs `alacritty -e /bin/sh -c 'htop' Htop`, which runs the commandline `htop` (the command `htop` with no further arguments) and sets `$0` of the shell to `Htop`.  Is this intended behavior?  If yes, I think documenting this effect of `-I` might be a good idea.  Also --dmenu respects `$SHELL` but this does not, as far as I can tell without looking at the source.

> Terminal emulators run in custom term mode are not run by a shell.

Aha!  I had suspected this, but apparently skipped over that part of the man page.  Thank you for telling me and for the suggestions for solving my original problem.

**Comment by meator** [comment]
> If I specify `-I` and launch Htop, it runs `alacritty -e /bin/sh -c 'htop' Htop`, which runs the commandline `htop` (the command `htop` with no further arguments) and sets `$0` of the shell to `Htop`. Is this intended behavior?

No, this isn't intended behavior.

I have "fixed" it in 2880efa0f993efe9f548a2eb5329381ac79639d8. The bug is still there, but it's now only reachable when j4-dmenu-desktop executes a terminal application which has `Path` set.

This heavily limits the desktop files which would be affected by this. Also, you appear to be aware of this, but I should reiterate that `--term 'alacritty -e {cmdline@} {name}'` does not make sense. Because of this, I don't plan to fix this in the near future. The command line assembly mechanism is complicated enough as is, this change would require additional changes which would complicate it further.

Do not hesitate to report any other custom term bugs though! Bug reports are welcome.

> Also --dmenu respects `$SHELL` but this does not, as far as I can tell without looking at the source.

This is intentional. j4-dmenu-desktop tries to use `/bin/sh` instead of `$SHELL` wherever it can. `/bin/sh` should be accessible on every POSIX compliant system and it is standardized. `$SHELL` can be anything. It may point to a non-POSIX compatible shell. `$SHELL` is used in `--dmenu` to enable dirty tricks [like this](https://github.com/enkore/j4-dmenu-desktop?tab=readme-ov-file#i-want-it-to-display-normal-binaries-too-yes). I have talked about this exceptional behavior in an older issue. It would probably make more sense to use `/bin/sh` for `--dmenu` too, but that would be a breaking change. But I still may chose to implement this breaking change in the future.

> Aha! I had suspected this, but apparently skipped over that part of the man page.

This isn't really documented. If your usage is so advanced that it would matter whether `--term` in custom term mode is run through `/bin/sh` or not, you should seriously consider making a separate script and calling it from `--term`, because it's simpler and it's behavior is clearly defined (the script interpreter is defined in the shebang line).

But I still intend to support custom term mode and fix its bugs.

I've documented this behavior now: c5786fc1d000906d7a9c045176b3028cd575110b
